services/src/main/scala/cromwell/pipeline/service/ProjectSearchService.scala

```diff
@@ -0,0 +1,115 @@
+package cromwell.pipeline.service
+
+import cromwell.pipeline.datastorage.dto._
+import cromwell.pipeline.model.wrapper.UserId
+import cromwell.pipeline.service.ProjectConfigurationService.Exceptions.NotFound
+import cromwell.pipeline.service.ProjectSearchService.Exceptions._
+import cromwell.pipeline.service.exceptions.ServiceException
+
+import scala.concurrent.{ ExecutionContext, Future }
+import scala.util.{ Failure, Success }
+
+trait ProjectSearchService {
+
+  def searchProjects(request: ProjectSearchRequest, userId: UserId): Future[Seq[Project]]
+
+}
+
+object ProjectSearchService {
+
+  object Exceptions {
+    sealed abstract class ProjectSearchException(message: String) extends ServiceException(message)
+    final case class InternalError(message: String = "Internal error") extends ProjectSearchException(message)
+  }
+
+  def apply(
+    projectService: ProjectService,
+    projectFileService: ProjectFileService,
+    projectConfigurationService: ProjectConfigurationService
+  )(
+    implicit ec: ExecutionContext
+  ): ProjectSearchService =
+    new ProjectSearchService {
+
+      override def searchProjects(request: ProjectSearchRequest, userId: UserId): Future[Seq[Project]] = {
+        val userProjects = projectService.getUserProjects(userId).recoverWith {
+          case _ => internalError("fetch projects")
+        }
+
+        def doSearch(searchFilter: ProjectSearchFilter): Future[Seq[Project]] = searchFilter match {
+          case AllFilter                    => userProjects
+          case ByNameFilter(mode, value)    => searchWithNameFilter(mode, value, userProjects)
+          case ByFilesFilter(_, hasFiles)   => searchByFilesFilter(hasFiles, userProjects, userId)
+          case ByConfigFilter(_, hasConfig) => searchByConfigFilter(hasConfig, userProjects, userId)
+          case ComplexOrFilter(leftFilter, rightFilter) =>
+            doSearch(leftFilter).flatMap(
+              resultLeft => doSearch(rightFilter).map(resultRight => resultLeft ++ resultRight)
+            )
+          case ComplexAndFilter(leftFilter, rightFilter) =>
+            doSearch(leftFilter).flatMap(
+              resultLeft => doSearch(rightFilter).map(resultRight => resultLeft.intersect(resultRight))
+            )
+        }
+        doSearch(request.filter).map(_.distinct)
+      }
+
+      private def searchByConfigFilter(
+        hasConfig: Boolean,
+        eventualProjects: Future[Seq[Project]],
```

Comment:
Methods accepting a `Future` are often signatures that the code "smells". Do these methods really need a `Future[Seq[Project]]` as an input? What's wrong with just `projects: Seq[Project]`?

Reply:
Refactored